### ISSUE TITLE
Moving external dependencies out of main repository?

### ISSUE THREAD

**Issue by dogtopus** [fabnicol/dvda-author#5]
I understand that external library deps maybe required for say Windows builds, but having them all checked in causes unnecessary performance overhead when working with the repository and wastes disk spaces.

**Comment by fabnicol** [comment]
You are overall right, to some extent. 
However, there is one crucial 3rd-party package (dvdauthor) that needs a good deal of patching, and to an older version, as I never had the time to port the patch to current. The patch was unfortunately not inlined by the core dvdauthor maintainers. And until its maintainer accepted the patch some years ago, mkisofs had to be patched too. 
Having a complete dependency tree built locally is a good insurance policy against versioning issues and inconsistencies. With software that has lots of dependencies, bitrot caused by dependency versioning drift is a major maintenance issue. Having a tested set of OK versions somewhat alleviates maintenance costs in the long run. Note that this app has been alive with an average of 15 downloads a day for 16 years now, and regularly updated. I prioritized stability over ease of build. Had I made different choices, the software would probably now be dead.
This said, you can build dvda-author without the 3rd-party dependencies, but this is not advised.

**Comment by dogtopus** [comment]
> Having a tested set of OK versions somewhat alleviates maintenance costs in the long run.

But it's still possible to make them git submodules (or just add the official repositories as submodules at specific commit) and have the build system accept both the builtin libraries or external packages on a per-library basis. This is a method used by a lot of big C/C++ game console emulator projects (such as RPCS3 and Citra) and should significantly cut down the repository bloat (currently it's 3GB checked out and it still downloads a lot of tarballs before even start building!). This way you can even mention about the external deps being less supported but you still gave the choice to potential package maintainers.

**Comment by fabnicol** [comment]
Good point. I'll try the submodule trick over this summer. Did you manage to complete a full build? I may end up getting just dvdauthor and lplex inlined in the source tree, but if for instance ImageMagick compatibility goes astray - this happened at least once - you will be on your own.

**Comment by fabnicol** [comment]
> the repository bloat (currently it's 3GB checked out and it still downloads a lot of tarballs before even start building!). This way you can even mention about the external deps being less supported but you still gave the choice to potential package maintainers.

If you want to avoid downloading dependencies, just build using:   
`./configure --enable-all-builds`  
instead of:   
`./configure --enable-all-all`   
And if you want to spee things up on *nix, use:
`sudo make PARALLEL=-j5`
(or any number of cores compatible with your platform).
Remember to use `sudo` at make stage. This is caused by a `cdrtools` install bug.

**Comment by dogtopus** [comment]
> Remember to use `sudo` at make stage. This is caused by a `cdrtools` install bug.

Invoking make (except maybe `make install`) with sudo seems to be a bit excessive in any case. What kind of bug is this? If it just checks for root before installing it may be possible to use fakeroot instead. Also it may be better to move all the install pass to `make install` instead but I'm not very sure if it applies well to this codebase since I'm not really familiar with the build system used here.

**Comment by fabnicol** [comment]
> > Remember to use `sudo` at make stage. This is caused by a `cdrtools` install bug.
> 
> Invoking make (except maybe `make install`) with sudo seems to be a bit excessive in any case. What kind of bug is this? If it just checks for root before installing it may be possible to use fakeroot instead. Also it may be better to move all the install pass to `make install` instead but I'm not very sure if it applies well to this codebase since I'm not really familiar with the build system used here.

Yes, I acknowledge this is not ideal but unfortunately cdrtools (mkisofs, cdrecord) installs a few files under /etc even if the install target is local. I consider this a bug of this dependency install process and there is no other way round.
Moving the local install process to make install is not uniformly possible as dvda-author uses FLAC, SoX and libogg to build. Originally FLAC source code was inlined in the build process of the 2005 Chapman version. This was changed  into FLAC library package building and linking at dvda-author build time, so you still need a built FLAC  library available at this stage. There is an arcane way of forcing platform library linking rather than building the dependency source package but this escape hatch is unfortunately unmaintained and probably broken now.
